artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java

```diff
@@ -209,7 +235,16 @@ private PageCache readPage(long pageId,
          }
          page.open();
          final long startedReadPage = System.nanoTime();
-         List<PagedMessage> pgdMessages = page.read(storageManager);
+         List<PagedMessage> pgdMessages;
+         if (indexCache == null) {
+            pgdMessages = page.read(storageManager);
+         } else {
+            TreeMap<Integer, Integer> messageNumberToOffset = new TreeMap<>();
```

Comment:
If this has to stay, Can we use a primitive map variant.

Reply:
Or even could this be simply an array of int, where array size will be max messageid - min messageid, and then simply access the offset by int[messageid - min messageid]

Reply:
And.. is there a way to encapsulate this logic on the PageIndexCacheImpl? just for code organization?